modules/imgproc/test/test_thresh.cpp

```diff
@@ -123,18 +135,75 @@ double CV_ThreshTest::get_success_error_level( int /*test_case_idx*/, int /*i*/,
 void CV_ThreshTest::run_func()
 {
     cvThreshold( test_array[INPUT][0], test_array[OUTPUT][0],
-                 thresh_val, max_val, thresh_type );
+                 thresh_val, max_val, thresh_type | extra_type);
 }
 
 
+static double compute_otsu_thresh(const Mat& _src)
+{
+    int depth = _src.depth();
+    int width = _src.cols, height = _src.rows;
+    const int N = 65536;
+    cv::Mat hMat(1, N, CV_32S);
+    int i, j;
+    int* h = &hMat.at<int>(0);
+    double mu = 0, scale = 1./(width*height);
+    hMat = 0;
+    for(i = 0; i < height; ++i)
+    {
+        for(j = 0; j < width; ++j)
+        {
+          const ushort val  = depth == CV_16UC1 ? _src.at<ushort>(i, j) : _src.at<uchar>(i,j);
```

Comment:
Consider casting both parts to the same type explicitly.
```.diff
-const ushort val  = depth == CV_16UC1 ? _src.at<ushort>(i, j) : _src.at<uchar>(i,j);
+const int val  = depth == CV_16UC1 ? (int)_src.at<ushort>(i, j) : (int)_src.at<uchar>(i,j);
```

(int => unsigned should work too)